---
title: "Results from rna_seq_kallisto pipeline"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
params:
  design: "test"
  stat: "star_stats.tab"
  p: "0.1"
  contrast_file: "contrast.tab"
  sessionId: "name"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(dplyr)
library(ggplot2)
library(reshape2)
library(readr)
library(tximport)
contrast_file = params$contrast_file ## maybe remove as parameter
s2c_file <- params$design
s2c <- read.table(s2c_file, header = TRUE, stringsAsFactors=FALSE,sep=",")
```

# {.tabset}

## Report info

### Intro

This is the html report with the results from the nextflow pipeline rna_seq_kallisto. The report contains figures and links to results. It also contains information about the pipeline run, such as which samples used, the parameters given to the pipeline, the softwares (incl version) used etc. This is important information and critical for being able to reproduce the results. It also information needed for writing the methods section should you happen to want to publish your results. Therefor, it is strongly recommended to keep the results folder and it's files as they are. If you want to do any post processing of e.g. the gene-lists, it is recommended to clearly document where the original file came from as well as what has been done to the data in the post processing.    

**PLEASE NOTE** that the *sessionId* explained in the tab "Used parameters, softwares and packages" can be used to keep track of the which run generated a certain file.

The different tabs in this report should be self explanatory, if anything is confusing please ask me (in person or by [email](mailto:elin.axelsson@gmi.oeaw.ac.at))
 

## DESeq2 Results

```{r p-value, echo= FALSE}
pvalue <- params$p
```


### Description of analysis
Differential expression analysis was carried out on the counts from [Kallisto](https://pachterlab.github.io/kallisto/about) using [DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html). For the import of the transcript level data and summarization into gene level data the R package [tximport](https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html). A tab file with the resulting counts can be found [here](../kallisto_data/kallisto_counts.tab).

**For information about versions used please refer to the tab "Used parameters, softwares and packages".**  The samples are described under tab "Sample info".

### Scatterplot matrix
<img src="pairs.png" alt="pairs plot">

*Matrix plot of all against all samples:* Each dot is one gene, the x and y axes are the two samples. That is the plot in row *i*, column *j* ( [*i,j*] ) is the scatter plot of the sample in [*i,i*] and the sample in [*j,j*].  E.g. the plot in [2,1] has the sample from [1,1] on the x-axis and the sample from [2,2] on the y-axis. The upper part of the matrix shows the spearman correlation between the corresponding samples.



### PCA plot
<img src="pca.png" alt="pca plot">

*PCA plot of all samples*: The plot shows the two largest PC, samples that are close to each other are more similar than samples further apart. Ideally one would like to see replicates clustering closely together and different conditions being clearly separated.


### Pair-wise DE analysis

For each contrast defined in the contrast file, a pair-wise differential expression test is being carried out. For dispersion estimates all samples provided to the pipeline are used.

```{r, echo=FALSE,include=FALSE, results='asis',comment=''}
out = NULL
for ( i in dir("maplots")){
out = c(out, knit_child("deseq_contrast.Rmd"))
}
```

`r paste(out, collapse='\n')`


## Alignment and aligment stats


The DE analysis is carried out using the kallisto output, that using the idea of pseudo-alignment and not any "normal" alignment. However, an alignment is useful for e.g. genome browser visualization and also as a quality matrix, hence in parallel to the DE analysis a [STAR](https://academic.oup.com/bioinformatics/article/29/1/15/272537) alignment was carried out.

The bigwig file (useful for genome browser) for each sample is available [here](../bam_bw/).

### Alignment stats

Below is alignment statistics as given from the STAR software. Especially the percentage of uniquely mapped reads and the average mapped length may be of interest. For RNA-seq we expect a high number of uniquely mapped reads. We also like to see that the proportion of uniquely mapped reads is approximately the same in all samples. Ideally the average mapped length should also be similar in all samples.

```{r star_stat, message=FALSE}
stat <- params$stat
star_stat=read.table(stat,sep="\t",header=T,stringsAsFactors = F)
star_stat = mutate(star_stat,sample=sapply(strsplit(ID,"Log"),"[[",1) )
star_stat = star_stat %>% select (-c(ID,X))
kable(star_stat, caption="Alignment stats (STAR)")
s2c <- mutate(s2c, name = paste(condition,sample,sep="_"))
s2c <-select(s2c, sample = run_accession, condition, name)
star_stat = inner_join(star_stat,s2c)
star_stat = star_stat %>% select (-c(sample,condition))
stat_mat = melt(star_stat, id = c("name"))
```

The data in the table can also be visualized by barplots:

```{r barplot}
ggplot(stat_mat,aes(x=name,y=value,fill=name))+ geom_bar(stat="identity")+facet_wrap(~variable,scales = "free")+ guides(fill=FALSE)+
  theme(axis.title.x=element_blank(),axis.title.y=element_blank())

```



## Sample info {#s_info}

```{r arguments, echo=FALSE}
s2c <- read.table(s2c_file, header = TRUE, stringsAsFactors=FALSE,sep=",")
kable(s2c,caption ="samples file used")
```



## Used parameters, softwares and packages

### Nextflow sessionId
The *sessionId* is an unique identifier that is connected to a specific run of a specific pipeline. This identifier is included (as a comment after # on the first line) in csv files generated by the pipeline,  making it possible to connect a csv file with it's report and the whole set of parameters, softwares and packages used to generate the file.

```{r,comment=NA}
cat(params$sessionId)
```
All scripts used are available [here](../used_script/) and the Nextflow script itself is stored [here](../nextflow/rna_seq1.nf).


### Software and versions
Those are the softwares (inc. version) that were loaded by the pipeline. For more detailed information see the file [nextflow.config](../nextflow/nextflow.config). 

```{r engine='bash', comment=''}
cat "modules.txt"
```


### Nextflow parameters
```{r engine='bash', comment=''}
cat "param.txt"
```

### Arguments passed on to R
```{r engine='bash', comment=''}
cat "Rarguments.txt"
```

### SessionInfo from DEseq2 R session
```{r engine='bash', comment=''}
cat "sessionInfo_deseq2.txt"
```




